### ISSUE TITLE
any suggestion for data IO when training on large scale dataset (17M images)

### ISSUE THREAD

**Issue by twmht** [NVIDIA/DALI#2796]
Hi,

I am using pytorch but having huge IO cost problem when training large scale dataset (17M images).

Storing image in LMDB or other kv store brings a huge IO cost.

is there any suggestions for reducing IO when training on such large scale dataset?

**Comment by JanuszL** [comment]
Hi @twmht,
Have you tried other storage formats like RecordIO or TFRecord?
As I understand you are concerned about inefficient storage format that adds a considerable amount of metadata on top of your data, right?
RecordIO can be created using [this script](https://github.com/apache/incubator-mxnet/blob/master/tools/im2rec.py), for TFRecord [this script](https://github.com/kmonachopoulos/ImageNet-to-TFrecord/blob/master/build_imagenet_data.py) may work (although I have not tested it).

**Comment by twmht** [comment]
@JanuszL 

I am trying recordIO but not sure if it can combine with pytorch dataloader when enabling workers.

**Comment by JanuszL** [comment]
I think you don't need to use the PyTorch data loader and can use DALI directly. [This example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/frameworks/pytorch/pytorch-various-readers.html) shows how you can use different data formats with PyTorch.

**Comment by twmht** [comment]
@JanuszL 

cool. I'll try that.

**Comment by twmht** [comment]
@JanuszL 

It seems that Dali only supports MXIndexedRecordIO. Does Dali support MXRecordIO?

**Comment by JanuszL** [comment]
Hi,
You need to use [this script](https://github.com/apache/incubator-mxnet/blob/master/tools/rec2idx.py) to create an index file, so your data set looks like [this](https://github.com/NVIDIA/DALI_extra/tree/master/db/recordio).

**Comment by twmht** [comment]
@JanuszL 

some people suggest to use Sequential file (https://github.com/deepinsight/insightface/issues/1426).

For example, https://github.com/deepinsight/insightface/blob/master/recognition/ArcFace/image_iter.py#L338

They use `ImageRecordIter` as a sequential file scanner., but Dali seems does not support that.

**Comment by JanuszL** [comment]
Hi,

What I see is the [MXRecordIO](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/recordio.py#L36) and [MXIndexedRecordIO](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/recordio.py#L215) are using the same underlying code. The only difference is the presence of idx file that can be used to randomly seek the data set. DALI doesn't seek the RecordIO file for shuffling, it does it only once on the very beginning to jump to shard assigned to DALI instance, that is why DALI requires it.
[The mentioned script](https://github.com/apache/incubator-mxnet/blob/master/tools/im2rec.py) has `no-shuffle` option that should be set to `false` as DALI required data set to be initially preshufled as the data set is read sequentially to the prefetch buffer used for shufling (you want to avoid having in the first batch samples from the classes available at the beginning of the data set).
To sumarize, you don't have to worry about MXIndexedRecordIO and MXRecordIO performance difference when using DALI.

**Comment by twmht** [comment]
@JanuszL 

you helped me a lot. thank you.
